crypto/err/err_blocks.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/err.h>
+#include "err_locl.h"
+
+void ERR_new(void)
+{
+    ERR_STATE *es;
+
+    es = ERR_get_state();
+    if (es == NULL)
+        return;
+
+    /* Allocate a slot */
+    err_allocate(es);
+    err_clear(es, es->top);
+}
+
+void ERR_set_debug(const char *file, int line, const char *func)
+{
+    ERR_STATE *es;
+
+    es = ERR_get_state();
+    if (es == NULL)
+        return;
+
+    err_set_debug(es, es->top, file, line, func);
+}
+
+void ERR_set_error(int lib, int reason, const char *fmt, ...)
+{
+    va_list args;
+
+    va_start(args, fmt);
+    ERR_vset_error(lib, reason, fmt, args);
+    va_end(args);
+}
+
+void ERR_vset_error(int lib, int reason, const char *fmt, va_list args)
+{
+    ERR_STATE *es;
+    char *buf = NULL;
+    size_t buf_len = 0;
+
+    es = ERR_get_state();
+    if (es == NULL)
+        return;
+
+    if (fmt != NULL
+        && (buf = OPENSSL_malloc(ERR_MAX_DATA_SIZE)) != NULL) {
+        int printed_len = 0;
+
+        printed_len = BIO_vsnprintf(buf, ERR_MAX_DATA_SIZE, fmt, args);
```

Comment:
If there's already a buffer, you should try to re-use it.  And that means keeping the buffer and size in the error-state.

Reply:
I'll look into that.

Reply:
I did a quick look, and I can only conclude that I have no idea what you're talking about.  "keeping the buffer and size in the error state" doesn't mean anything, as *this processes additional data* (i.e. the same stuff that  `ERR_add_error_data` deals with)

Reply:
The ```err_data``` field might already point to memory, and that memory might be big enough.  Does that help?

Reply:
How do I know that it's big enough?

Reply:
The current modus operandi is that when a new slot gets allocated, all its data is cleared (i.e. the `err_data` field is freed).

It's quite possible that we could make that more gentle on allocation/deallocation, but I'm starting to feel that it's beyond what I'm willing to put into *this* PR.  You have me thinking, though...  maybe I'll do something about that, in a separate PR...  or well, PR welcome, if anyone feels inclined!

Reply:
> How do I know that it's big enough?

That is why I originally said "you need to add a size field" :)

Reply:
Yeah, but like I said, this issue is becoming bigger than this PR, so wont fix here.

Reply:
https://github.com/openssl/openssl/issues/9458